### ISSUE TITLE
Verify official Gentoo/MIPS stage3 archive compatibility

### ISSUE THREAD

**Issue by frno7** [frno7/gentoo-mipsr5900el#16]
I’m happy to say that [Gentoo/MIPS](https://packages.gentoo.org/maintainer/mips@gentoo.org) lead member [Matt Turner](https://packages.gentoo.org/maintainer/mattst88@gentoo.org) applied the [`-mfix-r5900`](https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=fb51a3a867e20f574bde3b929ec9ccfba6cc374b) GCC option for the [R5900 short loop erratum](https://github.com/frno7/linux/wiki/R5900-short-loop-erratum) to the [Gentoo Catalyst](https://archives.gentoo.org/gentoo-catalyst/message/ede28a47143c8320420bc5df38250582) release building tool, which will make official Gentoo/MIPS stage3 [downloads](https://www.gentoo.org/downloads/) compatible with the R5900 and the PlayStation 2.

Verify official Gentoo/MIPS stage3 archives on PlayStation 2 hardware, and with [QEMU/R5900](https://github.com/frno7/qemu), when they become available.

**Comment by immolo** [comment]
https://gentoo.osuosl.org/releases/mips/autobuilds/20220505T204659Z/

Stage3 is ready for testing

**Comment by frno7** [comment]
Nice! I did a quick inspection of both mipsel2 and mipsel3 using the [`r5900check`](https://github.com/frno7/linux/tree/ps2-v5.4/tools/r5900check) tool. It reported a few problems, but not significantly more than can be expected with having `-mfix-r5900` at the moment. As noted in the article on the [R5900 short loop erratum](https://github.com/frno7/linux/wiki/R5900-short-loop-erratum), it may be some stray `noreorder` directives that remain to be fixed in some libraries. Example:

```
mipsel3 % r5900check lib/libc.so.6
erratum shortloop path lib/libc.so.6
code 000241f8 -3 c222eb44
code 000241fc -2 24010000
code 00024200 -1 e221eb44
code 00024204  0 1020fffc
code 00024208  1 00200825
```

So it seems there’s a good chance they’ll work.

**Comment by immolo** [comment]
Sources are verified and after checking only the mipsel3 stage contains the fix we need.

This can be closed now.

**Comment by frno7** [comment]
> Sources are verified and after checking only the mipsel3 stage contains the fix we need.

Hasn’t mipsel2 been given the `-mfix-r5900` treatment by Gentoo/MIPS? That’d be a pity since the R5900 to my knowledge is designed/supposed to be compatible with the whole range MIPS I, II and III. If someone for whatever reason likes to use mipsel2 with the R5900, why deny them that?

**Comment by immolo** [comment]
In my opinion it goes against the Gentoo way of optimising for your hardware however I can bring this up next time I'm talking to them.

**Comment by frno7** [comment]
Thanks! Before the `-mfix-r5900` option, the R5900 was mutually incompatible with every other MIPS processor in existence. Now that we have the option, it’d be nice if the R5900 could be part of the mipsel2 family too, and not excluded for being nonoptimal. Sometimes people seek the broadest possible hardware compatibility for various projects, and the R5900 has been on the sideline for so many years. :-)

**Comment by immolo** [comment]
I've discussed the mipsel2 stages and got the message that they didn't see a benefit to us or them from adding this.

My suggestion is we leave this now and I'll come back to it after we have working glibc and musl builds so we can lower the barrier of entry to get more people involved to help.

**Comment by immolo** [comment]
> Ah, well... Indeed, the Musl patch would be nice to have in.

Briefly spoke about that too in passing, we can have that added Gentoo side once I've tested it and that in turn will get it added upstream faster if it hasn't happened before hand.

On a side note has anyone here submitted a ps2 patch for gnuconfig recently? They were going to remove it until I offered to run some tests to see if we still needed it.

**Comment by frno7** [comment]
> On a side note has anyone here submitted a ps2 patch for gnuconfig recently? They were going to remove it until I offered to run some tests to see if we still needed it.

I have no idea what that thing does. :-) Do you have a link to it?

**Comment by immolo** [comment]
[0001-add-ps2-targets-to-config.sub.patch](https://dpaste.com/8KMWDKLB6)

Well it's something pre us as it was submitted in 2008 by one of the PS2 Gentoo devs.

**Comment by immolo** [comment]
Possibly, Sam (One of the lead devs) seems to think it could be more important for us though and as he is rarely wrong I'll play it safe and test it on a live system with and without before it's removed.

I'll update with the Gentoo bugreport once this is complete.

**Comment by immolo** [comment]
gnuconfig is used for configure I've found.

I've done some tests though and this is no longer required for us :)

https://bugs.gentoo.org/851045

**Comment by frno7** [comment]
Thanks!
